libcnb/src/layer/handling.rs

```diff
@@ -0,0 +1,679 @@
+                            // deserializing. Issue: https://github.com/Malax/libcnb.rs/issues/146
+                            //
+                            // Even if the deserialization of LayerContentMetadata is fixed, it would
+                            // not contain the layer types as they're not restored by the CNB lifecycle.
+                            // We must call layer.types here to get the correct types for the layer.
+                            types: layer.types(),
+                            metadata: layer_data.content_metadata.metadata,
+                        },
+                    )?;
+
+                    // Reread the layer from disk to ensure the returned layer data accurately reflects
+                    // the state on disk after we messed with it.
+                    read_layer(&context.layers_dir, layer_name.as_ref())?
+                        .ok_or(HandleLayerError::UnexpectedMissingLayer)
+                        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
+                }
+            }
+        }
+        Err(ReadLayerError::LayerContentMetadataParseError(_)) => {
+            match read_layer::<GenericMetadata, _, _>(&context.layers_dir, layer_name.as_ref()) {
+                Ok(Some(generic_layer_data)) => {
+                    let metadata_migration_strategy = layer
+                        .migrate_incompatible_metadata(
+                            context,
+                            &generic_layer_data.content_metadata.metadata,
+                        )
+                        .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+                    match metadata_migration_strategy {
+                        MetadataMigration::RecreateLayer => {
+                            delete_layer(&context.layers_dir, layer_name.as_ref())?;
+                        }
+                        MetadataMigration::ReplaceMetadata(migrated_metadata) => {
+                            write_layer(
+                                &context.layers_dir,
+                                layer_name.as_ref(),
+                                generic_layer_data.env,
+                                LayerContentMetadata {
+                                    types: generic_layer_data.content_metadata.types,
+                                    metadata: migrated_metadata,
+                                },
+                            )?;
+                        }
+                    }
+
+                    handle_layer(context, layer_name, layer)
+                }
+                Ok(None) => Err(HandleLayerError::UnexpectedMissingLayer.into()),
+                Err(read_layer_error) => {
+                    Err(HandleLayerError::ReadLayerError(read_layer_error).into())
+                }
+            }
+        }
+        Err(read_layer_error) => Err(HandleLayerError::ReadLayerError(read_layer_error).into()),
+    }
+}
```

Comment:
There's a lot going on in this method, but I don't see any obvious ways to re-structure it. We could port the error logic and non-error logic to different methods which would remove one indentation in our code rocket, it might make it easier to reason about. Not critical though. It's understandable, just dense.

Reply:
If you're fine with the current code, I'd prefer to not change it for now.

Reply:
That's fine.